### PR TITLE
Add change type option to fundrawtransaction

### PR DESCRIPTION
**Pull request by promag**
Adds a new option `change_type` to `fundrawtransaction` RPC. This is useful to override the node `-changetype` argument.

The new option is exclusive to `changeAddress` option, setting both raises a RPC error.

See also #11403, #12119.

### CODE REVIEW

**Comment by promag** [comment]
Missing tests for the new option.

**Comment by instagibbs** [comment]
might be a good candidate for another coincontrol field?

concept ACK

**Comment by promag** [comment]
> might be a good candidate for another coincontrol field?

I thought the same, but coin control doesn't sound the best place to put it. Correct me if I'm wrong though.

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3015,6 +3015,7 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
                             "   {\n"
                             "     \"changeAddress\"          (string, optional, default pool address) The bitcoin address to receive the change\n"
                             "     \"changePosition\"         (numeric, optional, default random) The index of the change output\n"
+                            "     \"change_type\"            (string, optional) The address type to use. Only valid if changeAddress is unspecified. Options are \"legacy\", \"p2sh\", and \"bech32\". Default is set by -changetype.\n"
```
Comment (jonasschnelli) [inline src/wallet/rpcwallet.cpp]: nit: missing `,` after `)`

**Comment by jonasschnelli** [comment]
Concept ACK.
I agree with @instagibbs that it could/should be a CoinControl member.

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (promag) [reply src/wallet/rpcwallet.cpp]: Not sure what you mean.

**Comment by promag** [comment]
Thanks @instagibbs and @jonasschnelli, changed as requested.

**Comment by promag** [comment]
Reference #12177.

**Comment by MeshCollider** [comment]
Does this close #11134 which is in the 0.16 milestone?

**Comment by kallewoof** [comment]
Also see #11489

**Comment by MarcoFalke** [comment]
I am -0 on this one, I tend to like #12119 more.

**Comment by promag** [comment]
If change type is explicit defined then it should follow #12119 logic for change type. I can rebase with that pull. WDYT?

**Review by ryanofsky — COMMENTED** [review]
utACK 4c74a8ef93d75e605feece08b150cdbd7260ec82. I think it's nice to be adding the coincontrol/fundraw option, if only to make it explicit that unless you override it (or `changeAddress`), then the type of change output will depend on global options the wallet was started up with. Notes:

* This PR will conflict with #12119, but the changes should be compatible (as promag pointed out). I don't have an opinion on which PR should be merged first or rebased on the other.
* It would be good to add a simple test to exercise the new code.
* It could be good if other RPCs calling CreateTransaction/SendMoney got a change_type option.

**Comment by promag** [comment]
> I don't have an opinion on which PR should be merged first or rebased on the other.

From what I can tell #12119 changes the `OUTPUT_TYPE_NONE` meaning. For that reason I would merge this one first.

> It would be good to add a simple test to exercise the new code.

Will do.

**Comment by promag** [comment]
Rebase and added test.

File: src/wallet/coincontrol.h [hunk 2]
```diff
@@ -17,6 +17,7 @@ class CCoinControl
 {
 public:
     CTxDestination destChange;
+    OutputType change_type;
```
Comment (instagibbs) [inline src/wallet/coincontrol.h]: Probably should add some comments distinguishing use between this and destChange right above.

Had to look up why both are required. (this being useful for non-prescribed destination, but prescribed destination type only)

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3057,6 +3057,7 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
                             "   {\n"
                             "     \"changeAddress\"          (string, optional, default pool address) The bitcoin address to receive the change\n"
                             "     \"changePosition\"         (numeric, optional, default random) The index of the change output\n"
+                            "     \"change_type\"            (string, optional) The address type to use. Only valid if changeAddress is unspecified. Options are \"legacy\", \"p2sh-segwit\", and \"bech32\". Default is set by -changetype.\n"
```
Comment (instagibbs) [inline src/wallet/rpcwallet.cpp]: maybe try to avoid additional "address" language for change stuff. "The output type to use for change"

**Comment by ryanofsky** [comment]
Slightly tested (running & tweaking fundraw test script) ACK utACK 79985b052e8846c11a165ac9eabcc8767c32cbb6. Changes since last review were rebasing after #12119, adding test, adding change_type RPCTypeCheck, adding coincontrol comments.

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/12194/commits/79985b052e8846c11a165ac9eabcc8767c32cbb6

non-blocking requests

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (promag) [reply src/wallet/rpcwallet.cpp]: Done.

File: src/wallet/coincontrol.h (hunk 2 above)
Comment (promag) [reply src/wallet/coincontrol.h]: Done.

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (promag) [reply src/wallet/rpcwallet.cpp]: Also s/unspecified/not specified.

File: src/wallet/coincontrol.h [hunk 4]
```diff
@@ -16,7 +16,10 @@
 class CCoinControl
 {
 public:
+    //! Custom change destination, if not set an address is generated
     CTxDestination destChange;
+    //! Custom change type, ignored if destChange is set, defaults to g_change_type
```
Comment (jonasschnelli) [inline src/wallet/coincontrol.h]: nit: the term "change type" seems not precise enough... I would have used `change address type`.

File: src/wallet/coincontrol.h [hunk 5]
```diff
@@ -16,7 +16,10 @@
...
+    OutputType change_type;
```
Comment (jonasschnelli) [inline src/wallet/coincontrol.h]: nit: I guess this should haven been `m_change_type`?

**Pull request merged**